### DISCUSSION TITLE
Connecting to board ...  done, Uploading sketch ...  done, Flashing sketch ... Error flashing the sketch

### DISCUSSION THREAD

**Discussion by imskanti** [JAndrassy/ArduinoOTA#138] [Q&A]
Used: C:\Program Files (x86)\Arduino\libraries\Ethernet
 Not used: C:\Users\skant\OneDrive\Documents\Arduino\libraries\EthernetENC-master
Using library SPI at version 1.0 in folder: C:\Users\skant\AppData\Local\Arduino15\packages\arduino\hardware\avr\1.6.23\libraries\SPI 
Using library Ethernet at version 2.0.0 in folder: C:\Program Files (x86)\Arduino\libraries\Ethernet 
Using library ArduinoOTA-1.0.8 at version 1.0.8 in folder: C:\Users\skant\OneDrive\Documents\Arduino\libraries\ArduinoOTA-1.0.8 
"C:\\Users\\skant\\AppData\\Local\\Arduino15\\packages\\arduino\\tools\\avr-gcc\\5.4.0-atmel3.6.1-arduino2/bin/avr-size" -A "C:\\Users\\skant\\AppData\\Local\\Temp\\arduino_build_844574/OTA_GATEWAY_test.ino.elf"
Sketch uses 20610 bytes (7%) of program storage space. Maximum is 261120 bytes.
Global variables use 1418 bytes (17%) of dynamic memory, leaving 6774 bytes for local variables. Maximum is 8192 bytes.
C:\Users\skant\AppData\Local\Arduino15\packages\arduino\tools\arduinoOTA\1.2.1/bin/arduinoOTA -address 192.168.1.104 -port 65280 -username arduino -password password -sketch C:\Users\skant\AppData\Local\Temp\arduino_build_844574/OTA_GATEWAY_test.ino.bin -upload /sketch -b 
Connecting to board ...  done
Uploading sketch ...  done
Flashing sketch ... Error flashing the sketch
Error flashing the sketch

**Comment by JAndrassy** [comment]
OTEthernet example works?

**Reply by imskanti** [reply]
No its not

**Comment by JAndrassy** [comment]
do you use my_boards or MegaCore? did you burn the Optiboot bootloader?

**Reply by imskanti** [reply]
yes im using my_boards. burnt the Arduino mega 2560(optiboot) in my boards

**Comment by JAndrassy** [comment]
can you try to upload with 'fake programmer'?
https://github.com/jandrassy/ArduinoOTA#ota-upload-from-ide-without-network-port

**Reply by imskanti** [reply]
I HAVE TRIED BOTH OF THE METHODES.

**Reply by imskanti** [reply]
As per my understanding its a issue with the bootloader, because connecting to board ... done , uploading sketch ... done but its not flashing.

i have tried arduino mega bootloader as well as your bootloader(optiboot) in both condition its same problem.

**Comment by JAndrassy** [comment]
why did you try with the Mega bootloader? it doesn't make sense to try it with that. 
Are you sure the Optiboot is successfully flashed?
the response sent to upload tool is before the uploaded sketch is applied so there is no way to report an error with flashing. it is just the upload tool's generic message.

**Reply by imskanti** [reply]
im sure Optiboot is flashed. 
im using arduino as isp to upload the bootloader is there any problem?

**Reply by JAndrassy** [reply]
> im using arduino as isp to upload the bootloade

me too. it is ok

**Reply by imskanti** [reply]
please help im stuck in the situation

can you give me the step by step check points

**Comment by JAndrassy** [comment]
I've seen this error too in some case, which I can't remember what it was or it go away before I could determine it.. The error is strange because checking the upload tool's source code I see that it prints  "Error flashing the sketch" only once.

can you show me the upload with the 'fake programmer'?

**Reply by imskanti** [reply]
Sketch uses 17370 bytes (6%) of program storage space. Maximum is 261120 bytes.
Global variables use 1329 bytes (16%) of dynamic memory, leaving 6863 bytes for local variables. Maximum is 8192 bytes.
C:\Users\skant\AppData\Local\Arduino15\packages\arduino\tools\arduinoOTA\1.2.1/bin/arduinoOTA -address 192.168.1.104 -port 65280 -username arduino -password password -sketch C:\Users\skant\AppData\Local\Temp\arduino_build_447877/OTA_GATEWAY_test.ino.bin -upload /sketch -b 
Connecting to board ...  done
Uploading sketch ...  done
Flashing sketch ... Error flashing the sketch
Error flashing the sketch

**Comment by JAndrassy** [comment]
I would expect it to use the arduinoOTA tool in my_boards/tools since the platform.txt file has
`tools.arduinoOTA.cmd={runtime.hardware.path}/tools/arduinoOTA`

**Reply by imskanti** [reply]
can you please clarify i did'nt got your point.

**Comment by JAndrassy** [comment]
upload with fake programmer with current version of my_boards should use
C:\Users\skant\Documents\Arduino\hardware\my_boards\tools\arduinoOTA
but your has
C:\Users\skant\AppData\Local\Arduino15\packages\arduino\tools\arduinoOTA\1.2.1/bin/arduinoOTA

**Comment by Bordman-ger** [comment]
Hi , let me follow up here with this topic, I have same issues with my OTA/WIFI combination, 
sometimes its working, sometimes not... (using Programmer, Serial monitor is closed, no Web Server Access in parallel
`Der Sketch verwendet 42742 Bytes (16%) des Programmspeicherplatzes. Das Maximum sind 261120 Bytes.
Globale Variablen verwenden 2329 Bytes (28%) des dynamischen Speichers, 5863 Bytes für lokale Variablen verbleiben. Das Maximum sind 8192 Bytes.
D:\Projekte\Arduino\data\hardware\my_boards/tools/arduinoOTA.exe -address 192.168.17.100 -port 65280 -t 30 -username arduino -password password -sketch C:\Users\Uwe\AppData\Local\Temp\arduino_build_992806/Wetterstation14.ino.bin -upload /sketch -b 
Connecting to board ...  done
Uploading sketch ...  done
Flashing sketch ... Error flashing the sketch
Error flashing the sketch`

If I repeat it several times it is somewhen working again... 
I am using some interrupts, tried to deactivate them, but even worse... so deleted the change again
  cli();
  ArduinoOTA.poll();
  sei();

**Reply by JAndrassy** [reply]
Mega with Ethernet library?

**Reply by Bordman-ger** [reply]
WeMOS Mega + WiFi R3 ATmega2560 + ESP8266 with WIFIESPAT

**Reply by JAndrassy** [reply]
the SerialWiFiOTA example works?

**Reply by Bordman-ger** [reply]
Yes everything is working, also OTA Updates works sometimes... b

**Comment by Bordman-ger** [comment]
`Der Sketch verwendet 42738 Bytes (16%) des Programmspeicherplatzes. Das Maximum sind 261120 Bytes.
Globale Variablen verwenden 2329 Bytes (28%) des dynamischen Speichers, 5863 Bytes für lokale Variablen verbleiben. Das Maximum sind 8192 Bytes.
D:\Projekte\Arduino\data\hardware\my_boards/tools/arduinoOTA.exe -address 192.168.17.100 -port 65280 -t 30 -username arduino -password password -sketch C:\Users\Uwe\AppData\Local\Temp\arduino_build_992806/Wetterstation14.ino.bin -upload /sketch -b 
Connecting to board ...  done
Uploading sketch ...  done
Flashing sketch ...  done
OK
Sketch uploaded successfully`

**Reply by Bordman-ger** [reply]
Is there any relation to the Web Server? 
now I tried 3 Updates after each other they worked, than I connected to the Web server - closed the browswer again, next Update fails...

**Reply by Bordman-ger** [reply]
waiting two minutes, next update worked again...

**Reply by Bordman-ger** [reply]
No - maybe not related to Web Server, now 2 Updades worked, the thirth is failing again, no Browser acces to the web server in between... forth updated worked again... but I had also situations where I had tu use USB again....

so let me know if I can do some special testing...

**Reply by JAndrassy** [reply]
didn't we already solve this?
https://github.com/JAndrassy/WiFiEspAT/discussions/97

**Reply by Bordman-ger** [reply]
no,  we fixed  that both is working in general in parallel .  yes that is fixed...
but Time by Time its not working... fix is still implemented...

Shall I stop the Web server and check if issue is gone?  I will do ...

**Reply by Bordman-ger** [reply]
4 times it worked, 5th time is failing again, Web server is off... so seem to be an other issue? 
 // start WEB Server
//  server.begin();
and
 // webserver();

**Reply by JAndrassy** [reply]
how large is the sketch? what is the baud rate for AT? 
do you have verbose mode for upload set in IDE preferences? it should print the cause of the upload error.
btw. don't turn off interrupts. Serial will not work efficient without interrupts

**Reply by Bordman-ger** [reply]
Size: 38798 Bytes (14%)
#define AT_BAUD_RATE 250000
  Serial3.begin(AT_BAUD_RATE);
  WiFi.init(&Serial3); 

Verbose is active for Upload only:
![grafik](https://user-images.githubusercontent.com/72354455/211206464-c7b1fca9-7f66-4396-80a5-0e88b1322f9d.png)

About the Interrupts: I have set 4 different interupts and I am using Serial1 in parallel - might this be a problem? 
Wifi is connected to Serial3

**Comment by JAndrassy** [comment]
to debug  you can add in WiFiOTA.cpp to `sendHttpResponse`
```
  Serial.print("OTA upload response: ");
  Serial.println(status);
```

**Reply by Bordman-ger** [reply]
put it in here, but no change...

void WiFiOTAClass::sendHttpResponse(Client& client, int code, const char* status)
{
  Serial.print("OTA upload response: ");
  Serial.println(status);
  
  while (client.available()) {
    client.read();
  }

  client.print("HTTP/1.1 ");

After changing I did restart IDE... anything else needed?

**Reply by JAndrassy** [reply]
what does it print to Serial Monitor?

**Reply by Bordman-ger** [reply]
no output ... 
sorry didnt try  the serial monitor before because:
If you test OTA with the 'fake' programmer, close the Serial Monitor in IDE. With Serial Monitor open, IDE will reset the board after upload tool finishes. The board then doesn't finish to apply the update.

I have to run it with a fake programmer... any chance to enhance the log of the upload...
or to see the client output?
![grafik](https://user-images.githubusercontent.com/72354455/211219884-99b9daa8-0823-4612-8ecf-49efcc881a42.png)

**Reply by JAndrassy** [reply]
does the upload tool (arduinoOTA) time out (30 seconds)?

**Reply by Bordman-ger** [reply]
yes it looks like this is a timeout
I was also able to run the update with this output after exactly 30 seconds...
Connecting to board ...  done
Uploading sketch ...  done
Flashing sketch ...  done
Error flashing the sketch
Error flashing the sketch

The update takes fifferent lenght, 20/25/17/30/19/26/18/24 20/15/27/25/30 seconds after last output...

so two times it timed out... and i got the error message this will also explain why its failing mor often if a client is connected to the Web server in parallel

**Reply by JAndrassy** [reply]
has the esp8266 a good WiFi signal? 38798 bytes large bin shouldn't need 30 seconds

**Reply by Bordman-ger** [reply]
I think its good... -44dB
the first two steps take less than one second...
Connecting to board ... done
Uploading sketch ... done

I startet the stop watch after "Flashing sketch ..." is shown...

**Reply by JAndrassy** [reply]
try longer timeout
-t parameter in platform.local.txt

if the upload is stable and only needs more time, then it should help

**Comment by Bordman-ger** [comment]
too easy (-: but it looks much better now.. I will keep watching - also running Webserver in parallel is okay...

**Reply by Bordman-ger** [reply]
But Question: when is "upload"  Done or do you mean upload + flashing?  - and why is Flassing taking 30 seconds? bad board design?  (integrated ESP)???

**Reply by JAndrassy** [reply]
ArduinoOTA does flashing while it receives the uploaded data. so those are not two separate steps.

**Reply by Bordman-ger** [reply]
update: its now working fine... 
setup: Client connected to the Webserver, in parallel OTA Update is done with WIFI, -47 dB connection...
I had only one issue today: after updating okay, the board was not longer reachable, after reboot it was okay... so maybe a reset issue??? but only 5% of updates will fail in this way...
Yesterday I had the situation tha the RX led was flassing but board was also not reachable - not with OTA, not with web client...
after reset it worked fine again...
I will pusht the board to the attic now - WIFI will be worse, we will see...

And please add this two in one board to the "working boards" list... clone of:WeMOS Mega + WiFi R3 ATmega2560 + ESP8266 

and as promised I will work on a quick start guide...

**Reply by megacct** [reply]
Hi, I have the same board as @Bordman-ger and the same issue with large sketches. Could you provide more detail on setting an increased timeout? -t isn't a valid arg for my ArduinoOTA.exe (ver 1.3.0). Thanks

**Reply by JAndrassy** [reply]
> Hi, I have the same board as @Bordman-ger and the same issue with large sketches. Could you provide more detail on setting an increased timeout? -t isn't a valid arg for my ArduinoOTA.exe (ver 1.3.0). Thanks

my build of the arduinoOTA in [my_boards](https://github.com/JAndrassy/my_boards) already has the -t parameter

**Reply by megacct** [reply]
Thanks. Hadn't spotted the newer version. -t option with this new version of the tool sorted my problem.
